### ISSUE TITLE
new dataloader

### ISSUE THREAD

**Issue by jacksalici** [jacksalici/PytorchTemplate#5]
With the current implementation of the SortDataset class ([link](https://github.com/jacksalici/PytorchTemplate/blob/8df6137f3fe576cef2b6eb41bd2985a1f8963cd7/dataloaders/toy_dataloader.py#L6C7-L6C18)), the samples returned by the `test_loader` could also be (all or partially) returned by the `train_loader`.

**Comment by FabioLanzi** [comment]
The [current solution](https://github.com/jacksalici/PytorchTemplate/commit/30989c82413f29eda6b61b385e385c6e84f73e83) isn’t the most elegant, but it's the one that requires the fewest changes.

I need to run tests with non-default parameters before moving this to `Done`. This is because the default parameters don't allow for enough unique sequences (I can only get 729 unique sequences of length 6 using only 3 digits).

**Comment by FabioLanzi** [comment]
Oops, I hadn't noticed. I'd say I ended up with a pretty similar solution, though.

**Comment by FabioLanzi** [comment]
I tested my fix (see [5-new-dataloader](https://github.com/jacksalici/PytorchTemplate/tree/5-new-dataloader)) by changing some default parameters. In particular, I increased the length of the training set, which now contains 12K unique examples by default (with `num_digits=5`).

Everything seems to be working fine and the model quickly reaches 100% accuracy.


```bash
python run.py --config default.yaml toySorter.yaml --task training
```

```
2025-09-03 14:18:52,639 [INFO] | Configuration: {
    "seed": 42,
    "force_reproducibility": true,
    "device": "cpu",
    "lr": 0.0005,
    "batch_size": 64,
    "num_epochs": 8,
    "model_name": "transformer",
    "input_dim": 10,
    "embed_dim": 32,
    "seq_len": 6,
    "num_layers": 3,
    "hidden_dims": null,
    "dropout_rate": 0.1,
    "activation": "relu",
    "use_batch_norm": true,
    "use_layer_norm": false,
    "output_dim": 2,
    "num_heads": 1,
    "attention_dropout": 0.1,
    "transformer_num_layers": 2,
    "experiment": "toysort",
    "dataloader": "toysort_ds",
    "optim": "adamw",
    "avoid_wandb": true,
    "save_model": true,
    "checkpoint_path": "checkpoints",
    "checkpoint_name": null,
    "config": [
        "default.yaml",
        "toySorter.yaml"
    ],
    "config_path": "configs",
    "project_name": "PytorchTemplate",
    "task": "training",
    "num_digits": 5
}
2025-09-03 14:18:53,405 [INFO] | Model: transformer, Parameters: 177,477
2025-09-03 14:18:55,421 [INFO] | {
    "Epoch": 0,
    "Train Loss": 0.8749436403022093,
    "Test Loss": 0,
    "Accuracy": 0.8109375,
    "Correct Sequences": 519,
    "Total Sequences": 640
}
2025-09-03 14:18:55,421 [INFO] | Saving model at epoch 0 with Test Loss 0
2025-09-03 14:18:56,618 [INFO] | {
    "Epoch": 1,
    "Train Loss": 0.3354236909093704,
    "Test Loss": 0,
    "Accuracy": 0.946875,
    "Correct Sequences": 606,
    "Total Sequences": 640
}
2025-09-03 14:18:56,618 [INFO] | Saving model at epoch 1 with Test Loss 0
2025-09-03 14:18:58,707 [INFO] | {
    "Epoch": 2,
    "Train Loss": 0.2314855980681863,
    "Test Loss": 0,
    "Accuracy": 0.990625,
    "Correct Sequences": 634,
    "Total Sequences": 640
}
2025-09-03 14:18:58,707 [INFO] | Saving model at epoch 2 with Test Loss 0
2025-09-03 14:19:00,495 [INFO] | {
    "Epoch": 3,
    "Train Loss": 0.19058891421652094,
    "Test Loss": 0,
    "Accuracy": 0.996875,
    "Correct Sequences": 638,
    "Total Sequences": 640
}
2025-09-03 14:19:00,495 [INFO] | Saving model at epoch 3 with Test Loss 0
2025-09-03 14:19:02,561 [INFO] | {
    "Epoch": 4,
    "Train Loss": 0.1704389418271136,
    "Test Loss": 0,
    "Accuracy": 1.0,
    "Correct Sequences": 640,
    "Total Sequences": 640
}
2025-09-03 14:19:02,561 [INFO] | Saving model at epoch 4 with Test Loss 0
2025-09-03 14:19:04,205 [INFO] | {
    "Epoch": 5,
    "Train Loss": 0.1534905042399697,
    "Test Loss": 0,
    "Accuracy": 0.9984375,
    "Correct Sequences": 639,
    "Total Sequences": 640
}
2025-09-03 14:19:04,205 [INFO] | Saving model at epoch 5 with Test Loss 0
2025-09-03 14:19:06,248 [INFO] | {
    "Epoch": 6,
    "Train Loss": 0.14636204320001092,
    "Test Loss": 0,
    "Accuracy": 1.0,
    "Correct Sequences": 640,
    "Total Sequences": 640
}
2025-09-03 14:19:06,248 [INFO] | Saving model at epoch 6 with Test Loss 0
2025-09-03 14:19:07,987 [INFO] | {
    "Epoch": 7,
    "Train Loss": 0.13524479697891736,
    "Test Loss": 0,
    "Accuracy": 1.0,
    "Correct Sequences": 640,
    "Total Sequences": 640
}
2025-09-03 14:19:07,987 [INFO] | Saving model at epoch 7 with Test Loss 0
```

**Comment by jacksalici** [comment]
Cool, ty! Feel free to open a PR when you want

**Comment by FabioLanzi** [comment]
I implemented an alternative solution (see: [-new-dataloader-alternative branch](https://github.com/jacksalici/PytorchTemplate/tree/5-new-dataloader-alternative), which moves sample generation from the `SortDataset` class initializer to the `get_dataloaders` function, based on @jacksalici’s suggestion.

This gives us two options that are stylistically different but functionally equivalent. Karpathy’s approach, on the other hand, is functionally different since it doesn’t provide control over duplicate sample generation within the same set.
